[WIP] Support implicit closing brackets

JMSS-Unknown opened the pull request:
Adds the functionality described in issue #133.
For example if you have
`1+2*3`,
you can add an opening bracket before the 1 and an implicit closing bracket will be placed at the end.
`(1+2*3)`
then, if you add a closing bracket after the two, the implicit bracket will disappear.
`(1+2)*3`

Current issues
* Does not work if nested - Should be a quick fix
* Does not work for expressions such as 1+2^2 that you want to turn into (1+2)^2 - Don't know how I'm going to add this
* Can't color one side of a bracket due to KaTeX issue https://github.com/KaTeX/KaTeX/issues/1844. Seems like there is a fix, that hasn't been merged https://github.com/KaTeX/KaTeX/pull/1845

saivan commented:
Just out of curiosity, how are you testing this?

JMSS-Unknown commented:
I haven't written any tests yet. So far, I just run the build script which generates build/guppy.js and just use the python simply server to run the example site locally.

saivan commented:
Yep, I do something similar :P was just checking - getting on it soon haha

saivan commented:
Okay thats freaking awesome :P I'd say the default character for right probably shouldn't be `\right|`, maybe `\color{#bbb}{\right}` or something to that effect? Or maybe it would be cool to implement a class instead (mathjax supports classes), which we can use for suggestions. That would be good instead of the question marked boxes as well because we could style suggestions as we like. @daniel3735928559 ?

JMSS-Unknown commented:
Yes, I want to make it \color{gray}\right)\color{black} but can't until https://github.com/KaTeX/KaTeX/pull/1845 gets merged. Also it can be styled as a custom class because @daniel3735928559 added an \xmlClass function https://github.com/KaTeX/KaTeX/pull/1791

saivan commented:
ahh, I don't know enough about xmlClass. but that is fair enough, seems like you have it covered. Seems awesome either way :) Great job!

JMSS-Unknown commented:
Thanks! :)

saivan commented:
Theres another question that this raises then (probably for another issue). If you entered `3x+4)` into your keyboard, then pressed `/`, should the brackets be consumed?

JMSS-Unknown commented:
It would definitely look better, so I think it should.

daniel3735928559 commented:
Functionally, I quite like how this works--though I haven't extensively tested it yet, the implicit closing bracket does about what I'd expect and e.g. resolves into an explicit closing bracket on its own pretty sensibly. Because there's a decent amount of code here, I'd like to think this through a little, but a few initial thoughts: 

* Maybe while waiting for KaTeX/KaTeX#1791 a reasonable stand-in might be something like `\right.\hspace{-.5ex}\color{#ccc}\text{)}`. The vertical bar as a closer is not very visually appealing

* Right now, four of the existing tests fail for me: `import_text calc`, `click unselect`, `click select left`, and click select right`.

* I don't quite understand what is happening (or really what is supposed to happen) when I enter `1 + 2 ) + 3 ) [left]`. Somehow, this takes me back to the beginning. My feeling is that the use of the second closing bracket should resolve the first guess opening bracket to a proper one and then create a new guess opening bracket

JMSS-Unknown commented:
I have addressed some of your points. Now typing in what you said results in
![image](https://user-images.githubusercontent.com/31131631/58777328-339cf080-8611-11e9-8791-c300fbd3575b.png)
Before, there was an error occurring clearing the editor. Another question is, should the inner guess bracket be resolved into a proper one?

In regards to the tests, I'll have a look at those now.

JMSS-Unknown commented:
The tests are failing because inserting `)` no longer moves out of the bracket when next to it. Now it creates a new guess open bracket resulting in `sin(())`. Is this behaviour acceptable? Then the tests need to be changed.

JMSS-Unknown commented:
Inner brackets are now converted to proper brackets automatically. Therefore typing what @daniel3735928559 said now results in
![image](https://user-images.githubusercontent.com/31131631/58789723-2ba17880-8632-11e9-8010-6fc6ca67037f.png)

saivan commented:
@daniel3735928559 can we merge this :P looks awesome!
